javalib/src/main/scala/java/util/Formatter.scala

```diff
@@ -590,101 +526,250 @@ final class Formatter private (private var dest: Appendable,
   }
 
   private def computerizedScientificNotation(
-      num: Number,
-      precision: Int,
+      x: Decimal,
+      digitsAfterDot: Int,
       forceDecimalSep: Boolean): String = {
 
-    val str = NumberFormatting
-      .formatScientific(num, precision)
-      .replace("E", "e+")
-      .replace("e+-", "e-")
+    val rounded = x.round(precision = 1 + digitsAfterDot)
 
-    // Finally, force the decimal separator, if requested
-    if (!forceDecimalSep || str.indexOf(decimalSeparator) >= 0) {
-      str
-    } else {
-      val pos = str.indexOf("e")
-      str.substring(0, pos) + decimalSeparator + str.substring(pos)
+    val signStr = if (rounded.negative) "-" else ""
+
+    val intStr               = rounded.unscaledValue
+    val dotPos               = 1
+    val fractionalDigitCount = intStr.length() - dotPos
+    val missingZeros         = digitsAfterDot - fractionalDigitCount
+
+    val significandStr = {
+      val integerPart    = intStr.substring(0, dotPos)
+      val fractionalPart = intStr.substring(dotPos) + strOfZeros(missingZeros)
+      if (fractionalPart == "" && !forceDecimalSep)
+        integerPart
+      else
+        integerPart + "." + fractionalPart
     }
+
+    val exponent        = fractionalDigitCount - rounded.scale
+    val exponentSign    = if (exponent < 0) "-" else "+"
+    val exponentAbsStr0 = Math.abs(exponent).toString()
+    val exponentAbsStr =
+      if (exponentAbsStr0.length() == 1) "0" + exponentAbsStr0
+      else exponentAbsStr0
+
+    signStr + significandStr + "e" + exponentSign + exponentAbsStr
```

Comment:
There's a lot of string concat in this method. Shouldn't we use StringBuilder instead? If yes, then we might try to pass it to `strOfZeros`